etl/scripts-ccao-data-warehouse-us-east-1/spatial/spatial-parcel.R

```diff
@@ -174,29 +194,137 @@ process_parcel_file <- function(s3_bucket_uri,
         bind_rows(spatial_df_missing)
     }
 
-    # Sort by year, town code, and PIN10 for better compression
-    spatial_df_merged <- spatial_df_merged %>%
+    # Convert to a data.table of the vertices of each polygon. Using data.table
+    # here because it's much faster
+    spatial_mat_coords <- spatial_df_merged %>%
+      st_set_geometry("geometry_3435") %>%
+      st_coordinates() %>%
+      as.data.table()
+
+    # Use the vertices to calculate additional features about the shape of
+    # the polygon
+    spatial_mat_calc <- spatial_mat_coords[
+      ,
+      `:=` (
+        # Get edge length using Pythagorean theorem and a rolling lag to get
+        # the distance between each pair of points
+        edge_len = sqrt(
+          (X - data.table::shift(X, type = "lag")) ^ 2 +
+            (Y - data.table::shift(Y, type = "lag")) ^ 2
+        ),
+        # Distance between the centroid and each point in the polygon
+        dist_to_centroid = sqrt(
+          (X - mean(X)) ^ 2 +
+            (Y - mean(Y)) ^ 2
+        ),
+        # Calculate the interior angle between each pair of points
+        angle = calculate_angles(as.matrix(.SD))
+      ),
+      .SDcols = c("X", "Y"),
+      by = c("L1", "L2", "L3")
+    ]
+
+    # Collapse the vertex-level data back to the parcel level EXCLUDING any
+    # vertices with an angle close to 180 degrees. These are likely to be
+    # artifacts of the data and not true vertices, so we don't count them in
+    # our calculations
+    spatial_mat_calc_vert <- spatial_mat_calc[
+      !data.table::between(angle, 179, 181) | is.na(angle),
+      # Number of vertices in the polygon. Minus 1 because the first and last
+      # vertex are always identical (used to close the polygon)
+      .(
+        shp_parcel_num_vertices = .N - 1,
+        # Tail is used to drop the first row of each group since it's identical
+        # to the last row (again, same reason as above)
+        shp_parcel_interior_angle_sd = sd(tail(angle, -1), na.rm = TRUE),
+        shp_parcel_centroid_dist_ft_sd = sd(tail(dist_to_centroid, -1))
+      ),
+      by = "L3"
+    ]
+
+    # Collapse the edge-level data back to the parcel level, this time keeping
+    # the 180 degree vertices, since excluding them would drop some edges with
+    # non-zero length
```

Comment:
**[Question, non-blocking]** Would it be slightly more accurate to sum the length of any edges that are joined with angle ~180° when executing operations based on edge length, since the additional vertex is most likely a data error cutting a single edge in two? I don't feel super strongly about this (especially if the universe of 180° vertices is small) but I figure it's worth checking!